src/main/scala/is/hail/expr/Type.scala

```diff
@@ -22,23 +22,28 @@ import scala.reflect.classTag
 abstract class BaseType
 
 object Type {
-  val genScalar: Gen[Type] = Gen.oneOf[Type](TBoolean, TInt32, TInt64, TFloat32, TFloat64, TString,
-    TVariant(GenomeReference.GRCh37), TAltAllele, TGenotype, TLocus(GenomeReference.GRCh37), TInterval(GenomeReference.GRCh37), TCall)
+  val genScalar: Gen[Type] = Gen.oneOf[Type](TBooleanOptional, TInt32Optional, TInt64Optional, TFloat32Optional, TFloat64Optional, TStringOptional,
+    TVariant(GenomeReference.GRCh37), TAltAlleleOptional, TGenotypeOptional, TLocus(GenomeReference.GRCh37), TInterval(GenomeReference.GRCh37), TCallOptional)
```

Comment:
It seems a little wrong that we never generate required scalar types?

Maybe have `genScalar`, `genRequiredScalar`, and `genOptionalScalar` and use them wherever is sensible? I guess `genScalar` could just be:
```scala
val genScalar = Gen.coin(0.2).flatMap(x => if (x) genRequriedScalar else genOptionalScalar)
```

Also all this proliferation of the prefix `gen` makes me think we should have a sub object called `gen` that contains all this stuff. Don't worry about that for this PR though. I've been thinking we generally need some way to manage Gen stuff better and more uniformly. Thoughts welcome!